runtime/src/installed_scheduler_pool.rs

```diff
@@ -391,11 +486,72 @@ impl BankWithScheduler {
     }
 
     pub const fn no_scheduler_available() -> InstalledSchedulerRwLock {
-        RwLock::new(None)
+        RwLock::new(SchedulerStatus::Unavailable)
     }
 }
 
 impl BankWithSchedulerInner {
+    fn with_active_scheduler(
+        self: &Arc<Self>,
+        f: impl FnOnce(&InstalledSchedulerBox) -> ScheduleResult,
+    ) -> ScheduleResult {
+        let scheduler = self.scheduler.read().unwrap();
+        match &*scheduler {
+            SchedulerStatus::Active(scheduler) => {
+                // This is the fast path, needing single read-lock most of time.
+                f(scheduler)
+            }
+            SchedulerStatus::Stale(_pool, (result, _timings)) if result.is_err() => {
+                Err(SchedulerAborted)
+            }
+            SchedulerStatus::Stale(_pool, _result_with_timings) => {
+                drop(scheduler);
+
+                let context = SchedulingContext::new(self.bank.clone());
+                let mut scheduler = self.scheduler.write().unwrap();
+                scheduler.transition_from_stale_to_active(|scheduler_pool, result_with_timings| {
+                    scheduler_pool.register_timeout_listener(self.do_create_timeout_listener());
+                    scheduler_pool.take_resumed_scheduler(context, result_with_timings)
+                });
+                drop(scheduler);
+
+                self.with_active_scheduler(f)
+            }
+            SchedulerStatus::Unavailable => panic!(),
+        }
+    }
+
+    fn do_create_timeout_listener(self: &Arc<Self>) -> TimeoutListener {
+        let weak_bank = Arc::downgrade(self);
+        TimeoutListener::new(move |scheduler_pool| {
+            let Some(bank) = weak_bank.upgrade() else {
+                return;
+            };
+
+            let Ok(mut scheduler) = bank.scheduler.write() else {
+                return;
+            };
+
+            scheduler.maybe_transition_from_active_to_stale(|scheduler| {
+                let (result_with_timings, uninstalled_scheduler) =
+                    scheduler.wait_for_termination(false);
+                uninstalled_scheduler.return_to_pool();
+                (scheduler_pool, result_with_timings)
+            })
```

Comment:
want to make sure I understand this correctly.
The closure here is called **before** we transition to Stale. So we wait for termination while the scheduler is still in active state.

I guess I'm not sure what happens here; it seems this intended to kill an extremely long running scheduler. But description of `wait_for_termination` says
```rust
    /// This function blocks the current thread while waiting for the scheduler to complete all of
    /// the executions for the scheduled transactions and to return the finalized
    /// `ResultWithTimings`. This function still blocks for short period of time even in the case
    /// of aborted schedulers to gracefully shutdown the scheduler (like thread joining).
```

So if it's extremely long-running, aren't we still blocked here? I assume I am missing something.

Reply:
> The closure here is called **before** we transition to Stale. So we wait for termination while the scheduler is still in active state.

yes. this is true.

> it seems this intended to kill an extremely long running scheduler.

well, this isn't true. the intention here is to gracefully terminate extremely long _idling_ scheduler. This is to reclaim the idling os native threads to return them to the scheduler pool.

> So if it's extremely long-running, aren't we still blocked here? I assume I am missing something.

this is correct. The situation is same both for blockstore processor and unified scheduler from the very beginnings.

Reply:
ah, also note that even while the scheduler constantly running with new transactions are arriving, this `wait_for_termiantion()` will eventually return. That's because we're taking the write lock and it will stop newer transactions are entered to the scheduler.